// @roots/bud-server/lib/server/watcher.js
import { __decorate, __metadata } from "tslib";
import chokidar from '@roots/bud-support/chokidar';
import { bind } from '@roots/bud-support/decorators';
/**
 * FS Watcher
 */
export class Watcher {
    /**
     * App instance
     */
    get app() {
        return this._app();
    }
    /**
     * Logger
     */
    get logger() {
        return this.app.server.logger.scope(`watcher`);
    }
    /**
     * Class constructor
     *
     * @param app - Application instance
     */
    constructor(_app) {
        this._app = _app;
    }
    /**
     * Watcher callback
     */
    watcherCallback(path) {
        this.logger.log(`edit to`, path.replace(this.app.path(), `.`), `triggered reload`);
        this.app.server.appliedMiddleware?.hot?.publish({
            action: `reload`,
            message: `Detected file change: ${path}. Reloading window.`,
        });
    }
    /**
     * Initialize watch files
     */
    async watch() {
        if (this.app.isCLI() && this.app.context.args.dry)
            return;
        this.files = this.app.hooks.filter(`dev.watch.files`, new Set([]));
        this.options = this.app.hooks.filter(`dev.watch.options`, {
            ignoreInitial: true,
            cwd: this.app.path(),
        });
        if (this.files.size < 1)
            return;
        this.instance = chokidar
            .watch([...this.files], this.options)
            .on(`change`, this.watcherCallback)
            .on(`add`, this.watcherCallback)
            .on(`unlink`, this.watcherCallback);
        this.logger.log(`watching`);
        return this.instance;
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", void 0)
], Watcher.prototype, "watcherCallback", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Watcher.prototype, "watch", null);
