// @roots/bud/lib/cli/commands/bud.build.production.js
import { Command } from '@roots/bud-support/clipanion';
import BuildCommand from './bud.build.js';
/**
 * `bud build production` command
 */
export default class BuildProductionCommand extends BuildCommand {
    constructor() {
        super(...arguments);
        /**
         * Subcommand context
         */
        this.withSubcommandContext = async (context) => {
            return {
                ...context,
                mode: `production`,
            };
        };
    }
}
/**
 * Command paths
 */
BuildProductionCommand.paths = [[`build`, `production`], [`production`]];
/**
 * Usage
 */
BuildProductionCommand.usage = Command.Usage({
    category: `build`,
    description: `Compiles source assets in \`production\` mode.`,
    details: `\
      \`bud build production\` compiles source assets in \`production\` mode.

      If you run this command without a bud configuration file \`bud\` will
      look for an entrypoint at \`@src/index.js\`.
    `,
    examples: [[`compile source assets`, `$0 build production`]],
});
