// @roots/bud-api/lib/methods/entry/guards.js
export function isPrimitive(input) {
    return (input.length === 1 &&
        (typeof input[0] === `string` || Array.isArray(input[0])));
}
export function isNamed(input) {
    return (input.length === 2 &&
        (typeof input[1] === `string` || Array.isArray(input[1])));
}
export function isNormalRecord(input) {
    return (input.length === 1 &&
        typeof input[0] !== `string` &&
        !Array.isArray(input[0]) &&
        Object.values(input[0]).every(value => typeof value !== `string` && !Array.isArray(value)));
}
export function isRecord(input) {
    return (input.length === 1 &&
        typeof input[0] !== `string` &&
        !Array.isArray(input[0]) &&
        Object.values(input[0]).every(value => typeof value === `string` || Array.isArray(value)));
}
export function isObjectWithArrayValues(input) {
    return Object.values(input).every(value => Array.isArray(value));
}
export function isObjectWithStringValues(input) {
    return Object.values(input).every(value => typeof value === `string`);
}
