// @roots/wordpress-theme-json-webpack-plugin/lib/plugin.js
import { __decorate, __metadata } from "tslib";
import fs from 'fs-jetpack';
import { bind } from 'helpful-decorators';
import omit from 'lodash/omit.js';
/**
 * ThemeJSONWebpackPlugin
 */
export class ThemeJsonWebpackPlugin {
    /**
     * theme.json path
     */
    get path() {
        return this.options.path;
    }
    /**
     * theme.json settings
     */
    get settings() {
        return JSON.stringify({
            __generated__: `⚠️ This file is generated. Do not edit.`,
            $schema: `https://schemas.wp.org/trunk/theme.json`,
            version: 2,
            ...omit(this.options, `path`),
        }, null, 2);
    }
    /**
     * Class constructor
     *
     * @param options - Plugin options
     */
    constructor(options) {
        this.options = options;
    }
    /**
     * Apply plugin
     *
     * @param compiler - Webpack compiler
     * @returns void
     */
    apply(compiler) {
        compiler.hooks.done.tapPromise(this.constructor.name, this.done);
    }
    /**
     * Compiler done callback
     *
     * @returns Promise
     */
    async done() {
        try {
            await fs.writeAsync(this.path, this.settings);
        }
        catch (err) {
            throw new Error(err);
        }
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", void 0)
], ThemeJsonWebpackPlugin.prototype, "apply", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], ThemeJsonWebpackPlugin.prototype, "done", null);
