// @roots/bud-support/lib/utilities/logger.js
import { __decorate, __metadata } from "tslib";
/* eslint-disable n/no-process-env */
import { bind } from 'helpful-decorators';
import Signale from 'signale';
import isUndefined from '../lodash/isUndefined/index.js';
import args from './args.js';
class Logger {
    constructor() {
        let options = {};
        if (args.log === false)
            options.disabled = true;
        options.logLevel = args.verbose ? `info` : args.log ? `log` : `warn`;
        if (process.env) {
            options.secrets = Object.entries(process.env)
                .filter((entry) => !isUndefined(entry[1]) && entry[0].includes(`SECRET`))
                .map(([k, v]) => v);
        }
        this.instance = new Signale.Signale(options);
        this.instance.config({ displayLabel: false });
    }
    log(...messages) {
        this.instance.log(...messages);
        return this;
    }
    time(label = `default`) {
        this.instance.time(label);
        return this;
    }
    timeEnd(label = `default`) {
        this.instance.timeEnd(label);
        return this;
    }
    success(...messages) {
        this.instance.success(...messages);
        return this;
    }
    info(...messages) {
        if (!(`verbose` in args))
            return this;
        this.instance.info(...messages);
        return this;
    }
    warn(...messages) {
        this.instance.warn(...messages);
        return this;
    }
    error(...messages) {
        this.instance.error(...messages);
        return this;
    }
    debug(...messages) {
        if (!(`verbose` in args))
            return this;
        this.instance.debug(...messages);
        return this;
    }
    await(...messages) {
        this.instance.await(...messages);
        return this;
    }
    scope(...scopes) {
        if (scopes.length === 0)
            return this;
        this.instance = this.instance.scope(...(scopes.filter(Boolean) ?? [`bud.js`]));
        return this;
    }
    unscope() {
        this.instance.unscope();
        return this;
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "log", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "time", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "timeEnd", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "success", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "info", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "warn", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "error", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "debug", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "await", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", void 0)
], Logger.prototype, "scope", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], Logger.prototype, "unscope", null);
let instance;
export default instance ?? (instance = new Logger());
export { Logger, instance };
