// @roots/bud-api/lib/api/service.js
import { __decorate, __metadata } from "tslib";
import { ServiceContainer } from '@roots/bud-framework/service';
import { bind } from '@roots/bud-support/decorators';
import { BudError, InputError } from '@roots/bud-support/errors';
import isFunction from '@roots/bud-support/lodash/isFunction';
import { factory } from '../facade/facade.factory.js';
import * as methods from '../methods/index.js';
/**
 * API service
 *
 * @remarks
 * Binds facade methods to the framework and provides a way to list them,
 * call them, and otherwise manipulate them.
 */
export class Api extends ServiceContainer {
    constructor() {
        super(...arguments);
        /**
         * Called methods
         */
        this.trace = [];
        /**
         * Queued method calls
         */
        this.queue = [];
    }
    /**
     * `bootstrap` callback
     */
    async bootstrap(_app) {
        Object.entries(methods).map(([k, v]) => this.bindFacade(k, v));
    }
    /**
     * Bind a synchronous facade for use in configs
     */
    bindFacade(name, fn) {
        if (!isFunction(fn)) {
            throw new BudError(`bud.api.bindFacade error: ${name} is not a function`);
        }
        this.set(name, fn.bind(this.app));
        this.app.bindMethod(name, factory(this.app, name));
    }
    /**
     * Call an api method directly
     */
    async call(name, args) {
        this.logger.info(`bud.api.call: ${name}`, ...args);
        if (!this.has(name)) {
            throw new InputError(`${name} is not a function`);
        }
        return await this.get(name).call(this.app, ...args);
    }
    /**
     * Execute all queued method calls
     */
    async processQueue() {
        const stack = [...this.queue];
        this.queue = [];
        await Promise.all(stack.map(async (value) => {
            if (!value) {
                this.logger.warn(`api.processQueue: undefined api call`);
                return;
            }
            const [name, args] = value;
            await this.call(name, args);
            this.trace.push([name, args]);
        }));
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Promise)
], Api.prototype, "bootstrap", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object]),
    __metadata("design:returntype", void 0)
], Api.prototype, "bindFacade", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Array]),
    __metadata("design:returntype", Promise)
], Api.prototype, "call", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Api.prototype, "processQueue", null);
