// @roots/wordpress-dependencies-webpack-plugin/lib/plugin.js
import { __decorate, __metadata } from "tslib";
import { bind } from 'helpful-decorators';
import Webpack from 'webpack';
import * as wpPkgs from './packages.js';
export default class WordPressDependenciesWebpackPlugin {
    constructor(options) {
        this.plugin = {
            name: `WordPressDependenciesWebpackPlugin`,
            stage: Infinity,
        };
        this.manifest = {};
        this.usedDependencies = {};
        this.fileName = options?.fileName ?? `wordpress.json`;
    }
    /**
     */
    apply(compiler) {
        compiler.hooks.normalModuleFactory.tap(this.plugin.name, this.normalModuleFactory);
        compiler.hooks.thisCompilation.tap(this.plugin, compilation => {
            this.compilation = compilation;
            this.compilation.hooks.processAssets.tap({ ...this.plugin, additionalAssets: true }, this.processAssets);
        });
    }
    /**
     */
    normalModuleFactory(factory) {
        factory.hooks.beforeResolve.tap(this.plugin.name, ({ contextInfo, request }) => {
            const { issuer } = contextInfo;
            this.usedDependencies = {
                ...this.usedDependencies,
                [issuer ?? `unknown`]: [
                    ...(this.usedDependencies[issuer] ?? []),
                    request,
                ].filter(wpPkgs.isProvided),
            };
            return;
        });
    }
    /**
     */
    processAssets(assets) {
        this.compilation.entrypoints.forEach(entry => {
            this.manifest[entry.name] = this.manifest[entry.name] ?? new Set([]);
            for (const chunk of entry.chunks) {
                this.compilation.chunkGraph
                    .getChunkModules(chunk)
                    .forEach(({ userRequest, modules }) => {
                    this.usedDependencies[userRequest]
                        ?.map((request) => wpPkgs.transform(request).enqueue)
                        .forEach((request) => {
                        this.manifest[entry.name].add(request);
                    });
                    modules?.forEach(({ userRequest }) => {
                        this.usedDependencies[userRequest]
                            ?.map((request) => wpPkgs.transform(request).enqueue)
                            .forEach((request) => {
                            this.manifest[entry.name].add(request);
                        });
                    });
                });
            }
        });
        assets[this.fileName] = new Webpack.sources.RawSource(JSON.stringify(Object.entries(this.manifest).reduce((manifest, [key, value]) => ({
            ...manifest,
            [key]: [...value],
        }), {})), true);
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Webpack.Compiler]),
    __metadata("design:returntype", void 0)
], WordPressDependenciesWebpackPlugin.prototype, "apply", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], WordPressDependenciesWebpackPlugin.prototype, "normalModuleFactory", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], WordPressDependenciesWebpackPlugin.prototype, "processAssets", null);
