// @roots/blade-loader/lib/loaders/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extract = exports.make = exports.repository = void 0;
const tslib_1 = require("tslib");
const css = tslib_1.__importStar(require("./css.cjs"));
const js = tslib_1.__importStar(require("./js.cjs"));
const scss = tslib_1.__importStar(require("./scss.cjs"));
const ts = tslib_1.__importStar(require("./ts.cjs"));
const vue = tslib_1.__importStar(require("./vue.cjs"));
exports.repository = {
    css: { ...css, loader: require.resolve(`./css.cjs`) },
    scss: { ...scss, loader: require.resolve(`./scss.cjs`) },
    js: { ...js, loader: require.resolve(`./js.cjs`) },
    ts: { ...ts, loader: require.resolve(`./ts.cjs`) },
    vue: { ...vue, loader: require.resolve(`./vue.cjs`) },
};
const make = function ({ extension, loader }) {
    const context = this.context || this.rootContext;
    const request = `${this.resource}.${extension}!=!${loader}!${this.remainingRequest}`;
    return `import ${JSON.stringify(this.utils.contextify(context, request))};`;
};
exports.make = make;
const extract = (source, pattern) => [...source.matchAll(pattern)]
    .map(({ groups }) => groups?.content)
    .filter(Boolean)
    .join(`\n`);
exports.extract = extract;
